### ISSUE TITLE
What are the semantics of `SyncAttrs.isVolatile`?

### ISSUE THREAD

**Issue by kyouko-taiga** [scala-native/scala-native#3610]
[LLVM](https://llvm.org/docs/LangRef.html#volatile-memory-accesses) describes "volatile memory accesses" as follows (emphasis mine):

> Certain memory accesses, such as [load](https://llvm.org/docs/LangRef.html#i-load)’s, [store](https://llvm.org/docs/LangRef.html#i-store)’s, and [llvm.memcpy](https://llvm.org/docs/LangRef.html#int-memcpy)’s may be marked volatile. The optimizers must not change the number of volatile operations or change their order of execution relative to other volatile operations. The optimizers may change the order of volatile operations relative to non-volatile operations. **This is not Java’s “volatile” and has no cross-thread synchronization behavior.**

That lead me to wonder what were the semantics of `SyncAttrs.isVolatile`.

**Comment by WojciechMazur** [comment]
`SyncAttrs.isVolatile` maps directly to `LLVM IR` volatile attribute for `load`/ `store` instructions. The JVM volatile behaviour is archived in other measures described below.  Current usages of LLVM `volatile` for load/store operations seams to be redundant and I believe it's usages could be removed. 

Every Scala field marked with `@volatile` would be marked with `Attr.Volatile`, and every `final` field would be marked with `Attr.Final` in compiler plugin when generating NIR. Field load/store operations for fields marked with these attributes would use LLVM IR `atomic load/store` with one of available memory orders - SeqCst for volatile fields, Monotonic for final and unordered for other fields (without any annotation). These should be sufficient to mostly comply with Java Memory Model.

**Comment by armanbilge** [comment]
> Monotonic for final ... These should be sufficient to mostly comply with Java Memory Model.

Hmm, why is it that `monotonic` correspond to `final` fields? Here's what LLVM says:

> If an address is written monotonic-ally by one thread, and other threads monotonic-ally read that address repeatedly, the other threads must eventually see the write.

That doesn't seem strong enough to guarantee that all threads will _always_ see the value of a `final` field _after_ it has been initialized.

**Comment by WojciechMazur** [comment]
> That doesn't seem strong enough to guarantee that all threads will always see the value of a final field after it has been initialized. 

Right, I forget about other aspect. If class has final fields then at the end of constructor call we do introduce a release fence. 
https://github.com/scala-native/scala-native/blob/b04fe54ddf02166558fe2a2e840230cfc81cda66/tools/src/main/scala/scala/scalanative/codegen/Lower.scala#L198-L206 
I'm not sure but in such case, probably the load of final field should use acquire memory order. We need to remember that it would be used for all final fields which in Scala are all `val` members of a class.

**Comment by armanbilge** [comment]
Thanks for that!

> probably the load of final field should use acquire memory order

I agree. It seems to me that _writing_ final fields doesn't need any special semantic, because it will be published via the fence anyway. Meanwhile, reading final fields should probably use acquire. [This implementation seems to be what is described in this blog post.](https://shipilev.net/blog/2014/all-fields-are-final/#_implementation_support). In either case, I don't see how `monotonic` is relevant.

> We need to remember that it would be used for all final fields which in Scala are all `val` members of a class.

Are you mentioning this because of the potential performance penalties? So long as we are implementing the Java Memory Model, our hands are tied :) it's a discussion for another day, but I'd be very interested if in the future the Scala Native semantics can be configured to use a more relaxed memory model.
